added_examples with optional arguments for rref methods in reductions.py

mohitacecode opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #12006 

#### Brief description of what is fixed or changed
added examples with optional arguments
#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

Commit:
added_examples

mohitacecode commented:
Please suggest if more examples should be there

mohitacecode committed:
remove trailspace

jksuom commented:
CC @siefkenj, @oscarbenjamin

Commit:
solved bug

mohitacecode committed:
fixes error

oscarbenjamin commented on sympy/matrices/reductions.py:
```diff
@@ -313,6 +313,14 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
+    >>> m.rref(pivots=True)
+    (Matrix([
+    [1, 0],
+    [0, 1]]), (0, 1))
```
Since True is the default I don't think we need this example. It would be better just to clarify that True is the default in the parameters section.

mohitacecode replied:
ohhk I will remove it

oscarbenjamin commented:
If this is going to close the linked issue then it will need an example using each of the optional parameters to demonstrate why they might be useful. Probably if you search in the tests you can find examples of them being used.

sylee957 commented:
I don’t think that it’s a good documentation only to list some code. 

You should have one sentence explanations between each examples

Commit:
Merge branch 'issue_12006' of https://github.com/mohitacecode/sympy into issue_12006

mohitacecode commented:
please tell me what these parameters do `normalize_last=True, simplify=False` Maybe with some example. because I am not able to figure it out.

oscarbenjamin commented:
@mohitacecode you have requested a review from me above but you haven't addressed the comments above from both me and @sylee957

mohitacecode commented:
I was working on it. I forget to addressed it but I have done what you both have said.I am just experiencing trouble with the `normalize_last=True` parameter.

oscarbenjamin commented:
> I am just experiencing trouble with the `normalize_last=True` parameter.

This parameter does not seem to be documented anywhere (hence the issue). The only way to figure out what it does is to look through the code.

mohitacecode commented:
okay I will look through the code and try to understand it. :)

mohitacecode commented:
Now the docstring look like this.
Any suggestions are welcome. @oscarbenjamin @sylee957 

```
    >>> from sympy import Matrix
    >>> from sympy.abc import x
    >>> from sympy.simplify.simplify import signsimp
    >>> m = Matrix([[1, 2], [x, 1 - 1/x]])
    >>> m.rref()
    (Matrix([
    [1, 0],
    [0, 1]]), (0, 1))

    rref return reduced row-echelon form of matrix and indices of pivot vars.

    >>> rref_matrix, rref_pivots = m.rref()
    >>> rref_matrix
    Matrix([
    [1, 0],
    [0, 1]])
    >>> rref_pivots
    (0, 1)

    If pivots is ``False`` just the row-reduced matrix is returned. By default pivots is set to ``True``.

    >>> m.rref(pivots=False)
    Matrix([
    [1, 0],
    [0, 1]])

    If different iszerofunc is used such as ``lambda x: abs(x)<6e-15``. By default ``lambda x: x.is_zero`` is used as a iszerofunc.

    >>> m = Matrix([[0.9, -0.1, -0.2, 0],[-0.8, 0.9, -0.4, 0],[-0.1, -0.8, 0.6, 0]])
    >>> m.rref(iszerofunc=lambda x: abs(x)<6e-15)
    (Matrix([
    [1, 0, -0.301369863013699, 0],
    [0, 1, -0.712328767123288, 0],
    [0, 0, 0, 0]]), (0, 1))

    If dotprodsimp is set to ``True``. By default dotprodsimp is set to ``False``.

    >>> M = Matrix(3, 3, [x+i for i in range (9)])
    >>> M.rref(dotprodsimp=True)
    (Matrix([
    [1, 0, -1],
    [0, 1,  2],
    [0, 0,  0]]), (0, 1))

    If different simplify method is used. By default SymPy's ``simplify`` is used.

    >>> M.rref(simplify=signsimp)
    (Matrix([
    [1, 0, (-(x + 1)*(x*(x + 5) - (x + 2)*(x + 3)) + (x + 2)*(x*(x + 4) - (x + 1)*(x + 3)))/(x*(x*(x + 4) - (x + 1)*(x + 3)))],
    [0, 1, (x*(x + 5) - (x + 2)*(x + 3))/(x*(x + 4) - (x + 1)*(x + 3))],
    [0, 0, 0]]), (0, 1))

    If normalize_last is set to ``False``. By default normalize_last is set to ``True``.

    >>> M.rref(normalize_last=False)
    (Matrix([
    [1, 0, -(x + 1)*(x + 5 - (x + 2)*(x + 3)/x)/(x*(x + 4 - (x + 1)*(x + 3)/x)) + (x + 2)/x],
    [0, 1, (x + 5 - (x + 2)*(x + 3)/x)/(x + 4 - (x + 1)*(x + 3)/x)],
    [0, 0, 0]]), (0, 1))

```

oscarbenjamin commented:
That looks like an improvement. It's easier to view and comment on the changes if you commit and push them.

I think that defaults for parameter values should be explained in the parameters section.

sylee957 commented:
I usually don’t mention default parameters if they are explicit keyword arguments and not determined automatically.

mohitacecode commented:
> I think that defaults for parameter values should be explained in the parameters section.

Okay I will remove that because they are already explained in the parameters section.

mohitacecode commented:
> That looks like an improvement. It's easier to view and comment on the changes if you commit and push them.

Okay I will commit and push these changes.

Commit:
required changes

mohitacecode commented:
> > It's easier to view and comment on the changes if you commit and push them.

Please have a look now. @oscarbenjamin @sylee957

oscarbenjamin commented on sympy/matrices/reductions.py:
```diff
@@ -314,6 +318,47 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
+    [1, 0],
+    [0, 1]])
+
+    If different iszerofunc is used such as ``lambda x: abs(x)<6e-15``.
```
This isn't a sentence. What happens in this case? We need to use words so that a user can understand. The example doesn't explain it.

oscarbenjamin commented on sympy/matrices/reductions.py:
```diff
@@ -314,6 +318,47 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
+    If dotprodsimp is set to ``True``.
+
+    >>> M = Matrix(3, 3, [x+i for i in range (9)])
+    >>> M.rref(dotprodsimp=True)
```
I think we need an example that demonstrates the effect of dotprodsimp e.g. by comparing the result of  rref with rref(dotprodsimp=True). Also the sentence should explain what it does.

oscarbenjamin commented on sympy/matrices/reductions.py:
```diff
@@ -314,6 +318,47 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
+    [0, 1,  2],
+    [0, 0,  0]]), (0, 1))
+
+    If simplify is set to ``signsimp``.
```
Why signsimp? Does it actually make sense to use signsimp?

mohitacecode replied:
I found it in tests and thought It would be better to use this as simplifier.

oscarbenjamin commented on sympy/matrices/reductions.py:
```diff
@@ -314,6 +318,47 @@ def _rref(M, iszerofunc=_iszero, simplify=False, pivots=True,
+    >>> M.rref(normalize_last=False)
+    (Matrix([
+    [1, 0, -(x + 1)*(x + 5 - (x + 2)*(x + 3)/x)/(x*(x + 4 - (x + 1)*(x + 3)/x)) + (x + 2)/x],
+    [0, 1, (x + 5 - (x + 2)*(x + 3)/x)/(x + 4 - (x + 1)*(x + 3)/x)],
```
It isn't clear what I should be looking at in this example

mohitacecode commented:
Okay will make improvements suggested by you.
Thanks for pointing out :)

mohitacecode commented:
It is defined in parameters section that what each optional argument do.
Do we still need to define their behaviour inbetween examples.

sylee957 commented:
In my opinion, normalize_last or dotprodsimp don’t need any examples because they already have ambiguous behavior for users.

sylee957 commented:
I don’t expect them to be documented. You can just note that they change the behavior how things are internally computed, but they do not affect results much

mohitacecode commented:
> Please tell me what this parameters do `normalize_last=True` maybe with some example. because I am not able to figure it out. I think it only affect the internal algorithm not the final output.

I was also thinking the same thanks for conforming.

mohitacecode commented:
Sorry for late reply I have done the changes suggested by @oscarbenjamin but as @sylee957 suggested that there is no need to add example for `normalize_last or dotprodsimp` so I have removed it.
what do you think about it are these changes seems ok to you.

oscarbenjamin commented:
This has merge conflicts that will need to be resolved.

The examples in the docstring don't really explain or demonstrate what the parameters do.

mohitacecode commented:
> The examples in the docstring don't really explain or demonstrate what the parameters do.

I have done some more changes I will push it soon.But as suggested by sylee I have removed examples for `normalize_last or dotprodsimp` so I was asking that what's your thought here should I removed that or keep it in my commit?

mohitacecode commented:
> This has merge conflicts that will need to be resolved.

Yes I agree

oscarbenjamin commented:
I think it's fine to remove those parameters from the docstring

mohitacecode commented:
> I think it's fine to remove those parameters from the docstring

Okay I will remove these.

mohitacecode commented:
I am having some confusion regarding merge conflict.
I was using these command to solve it but there were many files which are getting changed.

```
>git fetch upstream
>git merge upstream/master
then solving the merge conflict in `sympy/matrices/reductions.py`.
```
Is there any other better way that I should follow.

oscarbenjamin commented:
It might be worth just starting a new PR from master.

mohitacecode commented:
Okay sure

mohitacecode commented:
closing this one
